[TextField] Add in support for multiline text fields

peteratticusberg opened the pull request:
Keeps the api consistent with the api from v0.x.

Are we good with this implementation? Happy to make changes if need to be.

oliviertassinari commented on src/TextField/TextField.js:
```diff
@@ -108,10 +123,14 @@ export default class TextField extends Component {
         <Input
           className={inputClassName}
-          value={value}
+          component={multiLine ? 'textarea' : 'input'}
```
I'm wondering, shouldn't it be the `Input` responsibility to render a `textarea`?

peteratticusberg replied:
yea I think that's right. I'll change this.

peteratticusberg replied:
When I started making this change I realized it didn't make as much sense as I thought :(

The <Input /> component takes a `component` prop, and it feels wrong to sometimes determine the rendered component via the `component` prop and other times determine the rendered component via the `multiLine` prop. We'd have to do something like this: 

`const InputComponent = multiLine ? 'textarea' : ComponentProp;`

Which feels like unexpected behavior to me. Still happy to change this over if you like, but at this point I'm leaning towards keeping it the way it is.

oliviertassinari replied:
I have two thoughs here:
1. The `Input` component is public and can be used on his own, hence all the multiline logic should be at the same place, in the `Input`. 
2. Exposing a component property means user can override it. Hence, it should always take precedence, no matter the `multiLine` state.

oliviertassinari commented on src/Input/Input.js:
```diff
@@ -67,6 +67,12 @@ export const styleSheet = createStyleSheet('MuiInput', (theme) => {
+      resize: 'none',
+      'line-height': 'inherit',
+      padding: '0px',
+      'margin-top': '12px',
```
That performs better.
```css
margin: 6px 0 0;
padding: 0 0 6px;
```

oliviertassinari replied:
Also we need to set `margin: 0` on the input for Safari. It's broken right now.

oliviertassinari replied:
![capture d ecran 2017-04-09 a 12 17 29](https://cloud.githubusercontent.com/assets/3165635/24836520/87a84a70-1d1e-11e7-87b1-74a7d7120061.png)

peteratticusberg replied:
I think all 12px of margin needs to be on top of the element otherwise the text will overlap with the label for the field when there's overflow. Happy to try it this way though and see if that works

peteratticusberg replied:
How should we go about fixing the issue with safari?

peteratticusberg replied:
Here's what it looks like with `margin: 6px 0 0` `padding: 0 0 6px`

![screen shot 2017-04-09 at 2 51 50 pm](https://cloud.githubusercontent.com/assets/3895824/24840058/1574bcca-1d34-11e7-8888-341ea4ec554b.png)

the over flow still shows up a bit under the label

oliviertassinari replied:
Yes indeed, isn't that better? :) 
The Safari style can be fixed by adding a margin 0 on the input.

oliviertassinari replied:
```js
marginTop: 12,
```

oliviertassinari commented on src/Input/Input.js:
```diff
@@ -67,6 +67,12 @@ export const styleSheet = createStyleSheet('MuiInput', (theme) => {
     },
+    multiLine: {
+      resize: 'none',
+      'line-height': 'inherit',
```
I'm wondering if we need the `line-height: 1` style on the input.

peteratticusberg replied:
If we don't include line-height here the text gets all crunched together.

Are you saying here though that `line-height: 1` would be better than `line-height: inherit`? If so let's change this over. I'm really not very familiar with html/css, I generally just kind of "do what works".

peteratticusberg replied:
So I tried this out. 

Here's what it looks like with inherit:

![screen shot 2017-04-09 at 2 47 20 pm](https://cloud.githubusercontent.com/assets/3895824/24840025/83b464de-1d33-11e7-95f7-1f4213be8d8e.png)

And here's what it looks like with 1: 

![screen shot 2017-04-09 at 2 47 31 pm](https://cloud.githubusercontent.com/assets/3895824/24840027/8aecdb1e-1d33-11e7-9859-66f4c0d76411.png)

Think it makes sense to stick with inherit for now

oliviertassinari replied:
I was suggesting removing the `line-height` style properties. On the input and here.

oliviertassinari replied:
No need for that style once you rebase the PR on `next`.

oliviertassinari commented on src/Input/Input.js:
```diff
@@ -67,6 +67,12 @@ export const styleSheet = createStyleSheet('MuiInput', (theme) => {
       },
     },
+    multiLine: {
+      resize: 'none',
```
We don't want to allow users to resize?

peteratticusberg replied:
I used [v0](http://www.material-ui.com/#/components/text-field) as the spec for this, which doesn't allow  manual resizing for multiline text fields. 

I should've used [the official spec](http://www.material-ui.com/#/components/text-field) though...

Which says multiline text fields should work like this: 

![multilinetext](https://cloud.githubusercontent.com/assets/3895824/24837829/9e1d16dc-1d0a-11e7-8ce5-1ccb7176cd6f.gif)

Note that the spec also talks about Text Areas which are different from multi-line text fields, and which I think we'll also want to implement in a separate PR. 

![screen shot 2017-04-09 at 9 56 52 am](https://cloud.githubusercontent.com/assets/3895824/24837841/e247599e-1d0a-11e7-8377-48dc1852f32a.png)

oliviertassinari replied:
You raise a good point but I think that the concern is different. Mine isn't about the spec but the accesibility and UX. What do we win by disabling that feature?

peteratticusberg replied:
If the users are allowed to resize the textbox it could result in unexpected things happening in the UI (the user could make it too wide, or too tall). Especially if we allow them to resize textareas horizontally, which I think would really be a break with convention. 

We could allow users to resize textareas vertically but it would cause text to be cutoff mid-line unless we added javascript in to snap to (line-height * lines) sizes. Either solution here doesn't feel particularly clean to me. 

My preference here would be to either a) not do this at all or b) handle this in a separate PR. Let me know though if you want me to add it in here and I can take care of it.

oliviertassinari commented on src/TextField/TextField.js:
```diff
@@ -18,6 +18,10 @@ export default class TextField extends Component {
      */
     className: PropTypes.string,
     /**
+     * The default value for the TextField
```
``Default value of the `Input` element.``

peteratticusberg replied:
good catch

oliviertassinari commented on src/Input/Input.js:
```diff
@@ -114,6 +124,10 @@ export default class Input extends Component {
     /**
+     * If true, a textarea element will be rendered.
+     */
+    multiLine: PropTypes.bool,
```
Should it be called `multiline`? cc @mbrookes

peteratticusberg replied:
Yea, I actually had it as `multiline` originally but I noticed that in v0 it was `multiLine` so I refactored it to keep the API consistent. "multi-line" is one word, not two, so I think `multiline` would be more correct here. 

I know v1 is already going to be introducing breaking changes, so maybe we just change it to `multiline` here?

oliviertassinari replied:
Breaking changes aren't a concern. That new property name sounds better.

mbrookes replied:
Sorry, missed the cc. FWIW, I agree.

oliviertassinari commented on src/Input/Input.js:
```diff
@@ -98,6 +104,10 @@ export default class Input extends Component {
       PropTypes.func,
     ]),
     /**
+     * The default value for the string
```
``Default value of the input element.``

peteratticusberg replied:
good catch

oliviertassinari commented on src/Input/Input.js:
```diff
@@ -280,6 +302,7 @@ export default class Input extends Component {
           onFocus={this.handleFocus}
           onChange={this.handleChange}
           disabled={disabled}
+          defaultValue={defaultValue}
```
I don't think that we need to explicit the property, it's already inside the `other` variable when not destructured.

peteratticusberg replied:
Yea, that's a good point. So everything was working find with regular text fields based on an `<input />` but the defaultValue wasn't getting set for `<textarea>`'s until I explicitly passed defaultValue through as prop. 

This probably isn't the best solution... Any ideas on how to improve this? Or why it might not have been working via the `other` object for textareas?

oliviertassinari replied:
I confirm, you don't need that line.

peteratticusberg commented:
of course, thanks for the comments, I'll push up some changes in a bit 👨‍💻

peteratticusberg commented:
Hey, so after I merged in the latest version of `next` here I can no longer get my dev environment working. I get an issue with the command `webpack --config webpack.dll.config.js`:

```
ERROR in dll lib
Module not found: Error: Can't resolve 'babel-runtime' in '/Users/pete/code/mui/docs'
 @ dll lib
```

You'd think this would be easy to fix -- just install the babel-runtime dependency -- but the babel runtime dependency is already installed. I tried adding it to the dependencies of the docs project and still I got the same error. I'm running OSX Sierra, can't imagine that's affecting anything given this in a node error but who knows. 

@oliviertassinari I know you added this recently, any ideas what the issue might be here? If not I can keep digging on this one.

peteratticusberg commented:
Alright so here's the issue: 

the [babel runtime](https://github.com/babel/babel/tree/master/packages/babel-runtime) dependency doesn't specify a `'main'` file in it's package.json and it also does not have an index.js file in it's root. The entry point is `core-js.js`. I was able to fix this by manually editing the package.json file in the babel-runtime dependency to include `'main': 'core-js.js'`. 

This is of course not the right solution to this problem. 

I'm not sure what the right solution is but I'd imagine that it involves changing what we're doing in webpack.dll.config.js. I wonder if we could change the way we're dealing with dependencies there. 

I'm also curious -- is it just me that's having this problem? Are you able to run `yarn start` on the latest version of next?

mbrookes commented:
> I'm also curious -- is it just me that's having this problem?

I'm also having problems with `webpack.dll.config.js`, but haven't had a chance to did in to them...

oliviertassinari commented:
> here I can no longer get my dev environment working

Did you run yarn on the root **and** the docs directory since the update of the dependencies? If you had, I will try reproducing the issue on my end, but that's working fine with me as I'm writing.

mbrookes commented:
> Did you run yarn on the root and the docs directory since the update of the dependencies?

 I had, however `sudo ifconfig lo0 alias 10.200.10.1/24` seemed to fix it for me. I didn't used to have to run that. Not sure what changed...

peteratticusberg commented:
created a separate issue for this [here](https://github.com/callemall/material-ui/issues/6564)

kybarg commented:
Can I expect it be merged any time soon?

peteratticusberg commented:
@KapJI something like this will likely be merged sometime within the next two weeks, though it's unlikely that this exact implementation of multiline text fields will get merged cause it's not to spec (see above)

oliviertassinari commented on src/Input/Input.js:
```diff
@@ -67,6 +67,12 @@ export const styleSheet = createStyleSheet('MuiInput', (theme) => {
+    multiLine: {
+      resize: 'none',
+      'line-height': 'inherit',
+      padding: '0px',
```
```js
padding: 0,
```

oliviertassinari commented on src/Input/Input.js:
```diff
@@ -67,6 +67,12 @@ export const styleSheet = createStyleSheet('MuiInput', (theme) => {
         appearance: 'none',
       },
     },
+    multiLine: {
```
`multiline`

oliviertassinari commented on docs/src/pages/component-demos/text-fields/TextFields.js:
```diff
@@ -56,8 +56,15 @@ export default class TextFields extends Component {
+        <TextField
+          id="multiLine"
+          label="MultiLine"
+          multiLine
```
`multiline`

oliviertassinari commented on src/Input/Input.js:
```diff
@@ -152,6 +170,7 @@ export default class Input extends Component {
     disabled: false,
     type: 'text',
     disableUnderline: false,
+    multiLine: false,
```
`multiline`

oliviertassinari commented on test/regressions/tests/TextField/TextFieldMultiLine.js:
```diff
@@ -0,0 +1,17 @@
+    <div>
+      <TextField
+        label="Foo"
+        multiLine
```
multiline

oliviertassinari commented on src/TextField/TextField.spec.js:
```diff
@@ -32,6 +32,11 @@ describe('<TextField />', () => {
 
+    it('should render an Input with component="textarea" when passed the multiLine prop', () => {
+      wrapper = shallow(<TextField multiLine />);
+      assert.strictEqual(wrapper.childAt(0).prop('component'), 'textarea');
```
I find the following more idiomatic, but do it as you prefer
```js
.props().component
```

oliviertassinari commented on src/Input/Input.js:
```diff
@@ -238,10 +257,12 @@ export default class Input extends Component {
       className: classNameProp,
       component: ComponentProp,
-      inputClassName: inputClassNameProp,
+      defaultValue,
```
No need for that line.

oliviertassinari commented on src/Input/Input.js:
```diff
@@ -134,6 +148,10 @@ export default class Input extends Component {
      */
     onFocus: PropTypes.func,
     /**
+     * Number of rows to display when multiLine option is set to true.
```
`multiline`

oliviertassinari commented on src/TextField/TextField.js:
```diff
@@ -46,6 +50,10 @@ export default class TextField extends Component {
      */
     labelClassName: PropTypes.string,
     /**
+     * If true, a textarea element will be rendered.
```
``If `true`,``

oliviertassinari commented on src/Input/Input.js:
```diff
@@ -114,6 +124,10 @@ export default class Input extends Component {
      */
     inputClassName: PropTypes.string,
     /**
+     * If true, a textarea element will be rendered.
```
``If `true`,``

oliviertassinari commented on src/Input/Input.js:
```diff
@@ -238,10 +257,12 @@ export default class Input extends Component {
       disableUnderline,
       error: errorProp,
+      inputClassName: inputClassNameProp,
+      multiLine,
```
`multiline`

oliviertassinari commented on src/Input/Input.spec.js:
```diff
@@ -33,6 +33,12 @@ describe('<Input />', () => {
+  it('should render an <textarea /> inside the div when passed component="textarea"', () => {
+    const wrapper = shallow(<Input component="textarea" />);
+    const input = wrapper.find('textarea');
+    assert.strictEqual(input.is('textarea'), true, 'should be a <input>');
```
`should be a <textarea>`

oliviertassinari commented:
> that this exact implementation of multiline text fields will get merged cause it's not to spec (see above)

@peteratticusberg  Even if we don't follow closely the specification, I think that it's going in the right direction. We have to start somewhere.

peteratticusberg commented:
So I was able to get some work in on this today but there are still some issues with this.

Basically, we've given <input>'s padding, which works great, but if we give <textarea>'s padding, it causes a fractional number of rows to show up on scroll (e.g. 3.2) like so: 

![screen shot 2017-04-17 at 4 18 26 pm](https://cloud.githubusercontent.com/assets/3895824/25103181/858e6b40-2389-11e7-8efb-212488fef2b8.png)

This doesn't look very good and I think it's worth putting the time in to fix it. 

One way to "fix" this is to not give <textarea>s any padding. But if we do that, we end up with something like this:

![screen shot 2017-04-17 at 4 18 26 pm](https://cloud.githubusercontent.com/assets/3895824/25103270/dacb9aba-2389-11e7-98d1-aa5fd28145f1.png)

which is not consistent with how all the other TextFields look (they have 6px of padding- top and bottom). 

The only way to have what we want, is to stop padding the <inputs> directly, and instead apply the padding to the wrapper. I think this is the way to go. When we do this however, it causes some other slight visual differences I'm currently working on resolving.

oliviertassinari commented on src/TextField/TextField.js:
```diff
@@ -19,6 +19,10 @@ export default class TextField extends Component {
      */
     className: PropTypes.string,
     /**
+     * The default value of the `Input` element
```
Missing dot at the end.
``The default value of the `Input` element.``

oliviertassinari commented on src/Input/Input.spec.js:
```diff
@@ -33,6 +33,12 @@ describe('<Input />', () => {
+  it('should render an <textarea /> inside the div when passed the multiline prop', () => {
+    const wrapper = shallow(<Input multiline />);
+    const input = wrapper.find('textarea');
+    assert.strictEqual(input.is('textarea'), true, 'should be a <input>');
```
`should be '<textarea />'`

peteratticusberg replied:
just tried this but the test fails for `input.is('<textarea />')` 😞

oliviertassinari commented on src/Input/Input.js:
```diff
@@ -99,6 +107,10 @@ export default class Input extends Component {
       PropTypes.func,
     ]),
     /**
+     * The default value for the Input
```
Not consistent:
``The default value of the `Input` element.``

oliviertassinari commented on src/Input/Input.js:
```diff
@@ -68,6 +68,14 @@ export const styleSheet = createStyleSheet('MuiInput', (theme) => {
         appearance: 'none',
       },
     },
+    multiline: {
```
I have tried with the following properties, that looks all good to me, we can kill `multiline-wrapper`.
```js
    multiline: {
      resize: 'none',
      padding: '0 0 6px',
      marginTop: 6,
    },
    'multiline-wrapper': {
    }
```

![capture d ecran 2017-04-18 a 20 28 12](https://cloud.githubusercontent.com/assets/3165635/25146806/927670f4-2475-11e7-9226-1d31694b566d.png)

peteratticusberg replied:
Ah, yea I think that makes sense, good call

oliviertassinari commented on docs/src/pages/component-demos/text-fields/TextFields.js:
```diff
@@ -56,8 +56,15 @@ export default class TextFields extends Component {
         />
+        <TextField
+          id="multiline"
+          label="MultiLine"
```
`Multiline`?

oliviertassinari commented on src/Input/Input.js:
```diff
@@ -26,7 +26,7 @@ export const styleSheet = createStyleSheet('MuiInput', (theme) => {
         backgroundColor: palette.primary.A200,
         left: 0,
-        bottom: -1,
+        bottom: 0,
```
I don't think that we need to change that value.

peteratticusberg replied:
Yea this was a good point. So I looked into this and because I changed the position of the underline div I had to set it to `-2` to get the regression tests passing.

oliviertassinari commented on test/regressions/tests/TextField/TextFieldMultiLine.js:
```diff
@@ -0,0 +1,17 @@
+import React from 'react';
+import TextField from 'material-ui/TextField';
+
+export default function TextFieldMultiLine() {
```
Should we rename the file to `TextFieldMultiline.js`?

peteratticusberg commented:
Sounds good 🤝

mbixby commented:
I copied over EnhancedTextarea from master as Input/Textarea. With a couple of additional dirty tricks it seems to work (dynamic height, height transition). Anybody wants to clean it up? 

https://github.com/mbixby/material-ui/tree/multiline-inputs

oliviertassinari commented:
@mbixby This PR is closed from being done, but containes less than the master version. I guess we could do it in two steps. Your porte seems to work fine at least 😄

![avr -26-2017 23-50-32](https://cloud.githubusercontent.com/assets/3165635/25458763/8db1da74-2adb-11e7-93dd-86a3db11e3ae.gif)

peteratticusberg commented:
@mbixby's implementation actually matches the spec whereas this PR just re-implements the functionality provided by v0 (which was not to spec). 

Provided there aren't any major issues with this implementation (I don't see any after a brief review) I propose we close this PR and merge in @mixby's implementation. I'm happy to take care of this. 

We also should add a TextArea component at some point which is what the functionality added by this PR was intended to approximate. I may be able to get to this sometime in the next few weeks. 

For reference, the material design specs for these things:
![multiline spec](https://cloud.githubusercontent.com/assets/3895824/25956639/47262f32-363a-11e7-9b2f-8eedcde146d6.gif)
![text area](https://cloud.githubusercontent.com/assets/3895824/25956779/a2849972-363a-11e7-8d50-e6b0bb4bb44f.gif)

peteratticusberg commented:
Need to get the regressions for this passing but:

I started working on @mbixby's branch and realized that there were a lot of things in _this_ branch that I wanted. So I ended up just pulling in the auto-resizing `TextArea` component @mbixby created into this branch which now has the ability to render both auto resizing text areas (per the material design spec) and "multiline" textfields like the ones from v0. 

I think preserving the v0 multiline textfield functionality has two advantages. 1) It'll ease the transition for users 2) it'll act as a hold over in `next` until we get proper textareas. I'm not opposed to removing the v0 multiline textfields from this PR if we think that's the right thing to do.

mbixby commented:
@peteratticusberg any ideas about how to remove the `setTimeout` call in `AutoResizingTextArea#handleChange`? Seems hackish.

On my original branch I ran into issues with input height not being increased with newline on some slower devices (Safari on iPhone SE, complex app).

peteratticusberg commented:
@mbixby yea that's a fair point. [This](https://github.com/callemall/material-ui/pull/6553/commits/ab8a3bf78b5fe76739bdd7fa0974c484a2e62ab6) removes it. I still don't feel great about this but I do think it's an improvement.

peteratticusberg commented:
@oliviertassinari I recorded two new baselines, one for a Multiline TextField and another for the TextField demos page. 

It looks like Argos does not recognize these new baselines and is still testing against the old ones. Have you seen this happen before? Any ideas how to fix it? `yarn test:regressions` passes locally

oliviertassinari commented:
@peteratticusberg Thanks for pushing that branch forward! I'm gonna have a look at it tomorrow :). Regarding argos, that's expected. We shouldn't have to generate baselines in the first place, accepting changes is simpler (only members with write access can do so).

oliviertassinari commented on docs/src/pages/component-demos/text-fields/TextFields.js:
```diff
@@ -58,6 +58,21 @@ class TextFields extends Component {
           type="password"
         />
+        <TextField
+          id="multiline flexible"
```
I don't think that it's a valid id. `id="multiline-flexible"`

peteratticusberg replied:
good catch

oliviertassinari commented on docs/src/pages/component-demos/text-fields/TextFields.js:
```diff
@@ -58,6 +58,21 @@ class TextFields extends Component {
+          className={classes.input}
+        />
+        <TextField
+          id="multiline static"
```
I don't think that it's a valid id. `id="multiline-static"`

oliviertassinari commented on src/Input/AutoResizingTextArea.js:
```diff
@@ -0,0 +1,204 @@
+      lineHeight: 'inherit',
+      border: 'none',
+      outline: 'none',
+      'background-color': 'rgba(0,0,0,0)',
```
```js
background: 'transparent';
```

oliviertassinari commented on src/Input/AutoResizingTextArea.js:
```diff
@@ -0,0 +1,204 @@
+    const shadow = this.shadow;
+    const singleLineShadow = this.singleLineShadow;
+
+    const hasNewValue = (newValue && newValue !== '');
```
```js
const hasNewValue = newValue && newValue !== '';
```

oliviertassinari commented on src/Input/AutoResizingTextArea.js:
```diff
@@ -0,0 +1,204 @@
+    }
+  }
+
+  handleResize = (event) => {
```
I think that we should debounce the events, for instance:
https://github.com/callemall/material-ui/blob/next/src/Tabs/Tabs.js#L163-L173

oliviertassinari commented on docs/src/pages/component-demos/text-fields/TextFields.js:
```diff
@@ -58,6 +58,21 @@ class TextFields extends Component {
+          label="MultiLine"
+          multiline
+          defaultValue="Default Value"
+          className={classes.input}
```
What about adding `rowsMax={6}`?

peteratticusberg replied:
good call

oliviertassinari commented on src/Input/AutoResizingTextArea.js:
```diff
@@ -0,0 +1,204 @@
+    this.syncHeightWithShadow(undefined, event);
+  };
+
+  getInputNode() {
```
I don't think that we need that function indirection

peteratticusberg replied:
yea I agree

oliviertassinari commented on src/Input/AutoResizingTextArea.js:
```diff
@@ -0,0 +1,204 @@
+    return this.input;
+  }
+
+  setValue(value) {
```
I don't think that we should expose such method

oliviertassinari commented on src/Input/AutoResizingTextArea.js:
```diff
@@ -0,0 +1,204 @@
+          rows={this.props.rows}
+          defaultValue={this.props.defaultValue}
+          readOnly
+          value={this.props.value}
```
I guess we should add `aria-hidden="true"`.

oliviertassinari commented on src/Input/AutoResizingTextArea.js:
```diff
@@ -0,0 +1,204 @@
+          tabIndex="-1"
+          rows={1}
+          readOnly
+          value={''}
```
I guess we should add `aria-hidden="true"`.

peteratticusberg replied:
good call

peteratticusberg commented on src/Input/AutoResizingTextArea.js:
```diff
@@ -0,0 +1,204 @@
+
+  componentWillMount() {
+    // <Input> expects the components it renders to respond to 'value'
+    this.value = this.props.defaultValue;
```
It's so that `Input` can check whether an `AutoResizingTextArea` is dirty: 
```
function isDirty(obj) {
  return obj && obj.value && obj.value.length > 0;
}
```

peteratticusberg commented on src/Input/AutoResizingTextArea.js:
```diff
@@ -0,0 +1,204 @@
+          ref={(c) => { this.singleLineShadow = c; }}
+          className={classnames(classes.shadow, classes.textarea)}
+          tabIndex="-1"
+          rows={1}
```
rows is a `PropTypes.number` -- do you think it'd be better to make a `PropTypes.string`?

peteratticusberg commented:
Thanks for the review! Will fix this stuff up. 

And that's weird, TextFieldMultiline.png should be present: https://github.com/peteratticusberg/material-ui/blob/3161ccab80b454250b3288eb7eadd4a42e850757/test/regressions/screenshots/chrome/regression-TextField/TextFieldMultiline.png

peteratticusberg commented on src/Input/AutoResizingTextArea.js:
```diff
@@ -0,0 +1,204 @@
+  return {
+    root: {
+      position: 'relative', // because the shadow has position: 'absolute',
+      'margin-bottom': '-4px', // this is an unfortunate hack
```
Awesome.

peteratticusberg commented:
Alright! Addressed all those comments. Had to push up a new baseline for the TextField demo page (the difference was changing the label from MultiLine --> Multiline) don't know if we need to let argo know about that. 

Let me know if there's anything else you think we need to address before we merge this in.

peteratticusberg commented:
Awesome. Thanks for the help with this!
